lib/cdo/aws/cloud_formation.rb

```diff
@@ -0,0 +1,158 @@
+      if CDO.chef_local_mode
+        RakeUtils.with_bundle_dir(cookbooks_dir) do
+          Tempfile.open('berks') do |tmp|
+            RakeUtils.bundle_exec 'berks', 'package', tmp.path
+            Aws::S3::Client.new(region: 'us-east-1').put_object(
+              bucket: 'cdo-dist',
+              key: "chef/#{RakeUtils.git_branch}.tar.gz",
+              body: tmp.read
+            )
+          end
+        end
+      end
+    end
+
+    def self.update_bootstrap_script
+      Aws::S3::Client.new(region: 'us-east-1').put_object(
+        bucket: 'cdo-dist',
+        key: 'chef/bootstrap.sh',
+        body: File.read(aws_dir('chef-bootstrap.sh'))
+      )
+    end
+
+    def self.wait_for_stack(action, start_time)
+      CDO.log.info "Stack #{action} requested, waiting for instance provisioning to complete..."
+      begin
+        cfn.wait_until("stack_#{action}_complete".to_sym, stack_name: STACK_NAME) do |w|
+          w.max_attempts = 360 # 1 hour
+          w.delay = 10
+          w.before_wait { print '.' }
+        end
+      rescue Aws::Waiters::Errors::FailureStateError
+        CDO.log.info "\nError on #{action}. Event log:"
+        events = []
+        cfn.describe_stack_events(stack_name: STACK_NAME).stack_events.each do |event|
+          events << event
+          break if event.timestamp < start_time
+        end
+        events.reject{|event|event.resource_status_reason.nil?}.sort_by(&:timestamp).each do |event|
+          puts "#{event.timestamp}- #{event.logical_resource_id} [#{event.resource_status}]: #{event.resource_status_reason}"
+        end
+        return
+      end
+      CDO.log.info "\nStack #{action} complete."
+    end
+
+    def self.delete
+      stack_exists = !!cfn.describe_stacks(stack_name: STACK_NAME) rescue false
+      if stack_exists
+        CDO.log.info "Shutting down #{STACK_NAME}..."
+        start_time = Time.now
+        cfn.delete_stack(stack_name: STACK_NAME)
+        wait_for_stack(:delete, start_time)
+      else
+        CDO.log.warn "Stack #{STACK_NAME} does not exist."
+      end
+    end
+
+    def self.create_or_update(cdn_enabled=false)
+      json_template = json_template(cdn_enabled)
+
+      update_certs
+      update_cookbooks
+      update_bootstrap_script
+
+      stack_exists = !!cfn.describe_stacks(stack_name: STACK_NAME) rescue false
+      action = stack_exists ? :update : :create
+      CDO.log.info "#{action} stack: #{STACK_NAME}..."
+      start_time = Time.now
+      updated_stack_id = cfn.method("#{action}_stack").call(
+        stack_name: STACK_NAME,
+        template_body: json_template,
+        capabilities: ['CAPABILITY_IAM']
+      ).stack_id
+      wait_for_stack(action, start_time)
+      CDO.log.info 'Outputs:'
+      cfn.describe_stacks(stack_name: updated_stack_id).stacks.first.outputs.each do |output|
+        CDO.log.info "#{output.output_key}: #{output.output_value}"
+      end
+    end
+
+    def self.validate(cdn_enabled=false)
```

Comment:
Could this or any of the following methods be made private if they are for internal use only?